File: python/n509_FibNum.py
# https://leetcode-cn.com/problems/fibonacci-number/
class Solution:
    def fib(self, N: int) -> int:
        if N == 0: return 0
        return list(self.calfib(N))[-1]

    def calfib(self, N):
        a, b = 0, 1
        while N > 0:
            yield b
            a, b = b, a+b
            N -= 1
            

if __name__ == "__main__":
    pass

File: cpp/n509_FibNum.cpp
// https://leetcode-cn.com/problems/fibonacci-number/
#include <iostream>
#include <vector>
using namespace std;

class Solution{
public:
    int fib(int N){
        if (N == 0) return 0;
        if (N == 1) return 1;
        return (fib(N - 1) + fib(N - 2));
    }
    
    int fib_a(int N){
        if (N <= 1) return N;
        int dp0 = 0;
        int dp1 = 1;
        int dp2 = 1;
        for (int i = 2; i < N; ++i){
            int tmp = dp1 + dp2;
            dp1 = dp2;
            dp2 = tmp;
        }
        return dp2;
    }
};

int main(){
    return 0;
}